### PR TITLE
 feat: Add macros which map from RPC method strings to the RPC traits 

### PR DESCRIPTION
**Pull request by Marwes**
Implements the mapping I hinted at in https://github.com/gluon-lang/languageserver-types/pull/29#issuecomment-347686297. The idea is that one can just grep for the actual method names that are called when searching for how/if a method is implemented instead of guessing at what a method's "type" are called.

Based on #30 

cc @Xanewok

### CODE REVIEW

**Comment by Marwes** [comment]
Usage of this (+ #29) in the gluon language server https://github.com/gluon-lang/gluon_language-server/pull/15/commits/797ffe9f1c0e47418719cf9c08b8e75e5b5ce208

File: src/notification.rs [hunk 1]
```diff
@@ -5,14 +5,40 @@ pub trait Notification {
...
+    ("initialized") => { $crate::notification::Initialized };
+    ("exit") => { $crate::notification::Exit };
+
+    ("window/showMessage") => { $crate::notification::ShowMessage };
+    ("window/logMessage") => { $crate::notification::LogMessage };
+
+    ("telemetry/event") => { $crate::notification::Event };
+
+    ("client/registerCapability") => { $crate::notification::RegistrationParams };
+    ("client/unregisterCapability") => { $crate::notification::UnregistrationParams };
```
Comment (Xanewok) [inline src/notification.rs]: Same with [`UnregisterCapability`](https://github.com/gluon-lang/languageserver-types/pull/31/files#diff-028ce44e20d576b20f8c3ffbbbd75987R122)

**Comment by Xanewok** [comment]
That's convenient! It'd be great to enforce exhaustive macro in case there's a new method to be added or if we could somehow derive the macro from the impls themselves (using another macro?), but I can't think of a clean way that would not make this unnecessarily more complex.

File: src/notification.rs [hunk 2]
```diff
@@ -5,14 +5,40 @@ pub trait Notification {
...
+    ("$/cancelRequest") => { $crate::notification::Cancel };
+    ("initialized") => { $crate::notification::Initialized };
+    ("exit") => { $crate::notification::Exit };
+
+    ("window/showMessage") => { $crate::notification::ShowMessage };
+    ("window/logMessage") => { $crate::notification::LogMessage };
+
+    ("telemetry/event") => { $crate::notification::Event };
+
+    ("client/registerCapability") => { $crate::notification::RegistrationParams };
```
[the remark this answers is not in the document]
Comment (Marwes) [reply src/notification.rs]: Oops, missed updating this after copy pasting from the branch I had.

**Comment by Marwes** [comment]
> That's convenient! It'd be great to enforce exhaustive macro in case there's a new method to be added or if we could somehow derive the macro from the impls themselves (using another macro?), but I can't think of a clean way that would not make this unnecessarily more complex.

Yeah me neither :/ . They are easily detected when one tries to use the macro at least and easy to add after that. The number of methods added to LSP in the future is likely ~10 (order of magnitude) so it is not worth spending much time on.

**Comment by Marwes** [comment]
Released as 0.18.0

**Pull request merged**